include/revng/ABI/AlignmentMap.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <compare>
+#include <cstdint>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/STLExtras.h"
+
+#include "revng/Support/Debug.h"
+
+namespace abi {
+
+struct Alignment {
+public:
+  uint64_t Size;
+  uint64_t PadTo;
+  uint64_t AlignAt;
+
+public:
+  constexpr auto operator==(const Alignment &A) const { return Size == A.Size; }
+  constexpr auto operator<=>(const Alignment &A) const {
+    return Size <=> A.Size;
+  }
+};
+
+class AlignmentMap {
```

Comment:
This can be a simple `std::map` or a `SortedVector`.
Once we will introduce user customizable ABIs, this kind of things will have to be removed in favor of something YAML-serializable.
Given this perspective, I'd avoid them altogether.

Reply:
It is a `SortedVector`. It's just immutable, so I stripped all the vector-y bits and made it an array. Once we need it to be runtime-defined, it should just be a trivial change of the underlying container.

Reply:
* You can achieve immutability with `const`.
* It's not about changing the underlying container, it's about the fact that the whole description of the ABI should be YAML-serializable and this data structure is not.

Reply:
1. But then it doesn't provide compile-time access.

2. It's an `ArrayRef` of structs with three ints. It's super trivial to move that to the TTG at any point.

Also consider that this is at most an interface, it's built in-place by passing anything `llvm::ArrayRef<Alignment>` compatible. So any kind of array, vector, `SortedVector`, whatever - it's implicitly constructible.

In essence, this is just a
```cpp
const Alignment &selectAlignment(llvm::ArrayRef<Alignment> Table, uint64_t Size);
```
but wrapped into a class because then the usage looks just that little bit cleaner.

Reply:
Compile time access is a non-goal. We're going to make this user-customizable, why keep pushing for compile-time access? There's no point. This is a recurring theme in ABI PRs.

ABI descriptions will be user-customziable so I'm against all efforts to make it more efficient at compile-time. There's zero extra complexity I'm willing to accept for a non-goal.

I'm at the point where I'm going to ask you to make the ABI description YAML serializable so we don't get into this kind of discussions again, but that's not a priority as of now.

I understand you want to make things clean and efficient, but, given the direction we want to go, this is a waste of energy, review time and complexity.

Reply:
Yeah, gotcha, sorry about that.

Reply:
This is gone now. It became `TypeSpecific` part of the new runtime no-longer-trait.